rfc: Alter column type

bobvawter opened the pull request:
This RFC describes how we will support altering column types in an online
fashion.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/24703)
<!-- Reviewable:end -->

knz commented:
Some early feedback posted here with agreement from the author. 

In random order of importance:
- we usually hard wrap the text of rfcs.
- evolutation -> evolution
- you're missing a conversion class: "Assisted", which is even more general than what you now call "General". I suggest renaming "General" to "Cast-based" and add "Assisted" next to that. The new section "Assisted" is when an explicit conversion expression is specified with USING. I think this will be a pretty common case actually.
- appyling -> applying (may I suggest enabling the spell check in your editor? no judgment here)
- you will get questions about your deferred/completed approach with jobs. It seems orthogonal to this RFC: the mechanism could be equally well applied to any of our other asynchronous schema changes. If you want to argue it is necessary for "alter table set type", then I'd recommend starting this work separately, and complete it before you resume work on this feature.
- Cloumns -> columns
- I think that any value written to the old type after the change has been initiated, must also be done synchronously to the shadow column. We wouldn't know how to pick them up during the backfill otherwise. (Same as adding a new index.)
- "A backfill process is run on each node" -> "A backfill process is distributed across all nodes that have data for the table, like we do for index backfills already"
- " Once the backfill is complete, the column and indexes are upgraded to `READ` status.  Any code that resolves a column or index name to an id should be prepared to look for a `READ`-mode shadow that has been activated. " - I have no idea what you're talking about here (neither what this is about nor why this is important). You should clarify, with examples.
- there's a missing entry in failure modes: Writes that occur during the backfill (after the change is initiated, but before the new type becomes active) are issued by the client app _under the assumption that the old type is active_. This means that from those clients' perspective it is OK to write any value valid for the old type *but some of these values may be invalid for the new type*. You have to outline what happens in that case.
- the final paragraph "Solving for the general case does have the potential to become a science project" doesn't make much sense to me. What are you trying to say? The project is complex, that's true, but so are other projects on CockroachDB. What are the things that require science which will not be completed when the work defined by the RFC is done?

couchand commented:
Pro-tips: Bors understands the `do-not-merge` label, and also you're gonna want to update the PR description before you merge.

bobvawter commented:
@knz, thank you for your near-instant feedback.  I have incorporated the changes that you have suggested.

@couchand Fixed.

knz commented:
The RFC also needs to talk about interleaved tables and column families. How does the proposal work with those cases?

---

Reviewed 1 of 1 files at r1.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24703#-:-L9qqjeZ3Z4ceR8_UWRd:bt08ft8)*
<!-- Sent from Reviewable.io -->

mjibson commented on docs/RFCS/20180411_alter_column_type.md:
```diff
@@ -0,0 +1,303 @@
+
+## Indexed columns
+
+Columns that are part of an index are supported, however columns that
```
Does this (not changing a PK column) apply to all classes? Even `trivial`?

mjibson commented on docs/RFCS/20180411_alter_column_type.md:
```diff
@@ -0,0 +1,303 @@
+Given that a type-change process may take an arbitrary and unpredictable
+amount of time to complete, we support an optional two-phase approach to
+making the schema change visible to SQL clients.  This allows the user
+to choose exactly when a (breaking) change to the table schema will
```
Can you define breaking change? It seems like it's anything not in the trivial class.

mjibson commented on docs/RFCS/20180411_alter_column_type.md:
```diff
@@ -0,0 +1,303 @@
+
+The `ALTER TABLE` command is first run with the `DEFERRED` option, which
+begins the type-conversion in the background and returns an associated
+job id.  At any point in time, but generally after the job has
```
What's the use of the job id? We don't return them anywhere else, and I might want to add them all at once instead of here so we can decide exactly when, why, and how we want to return job ids.

mjibson commented on docs/RFCS/20180411_alter_column_type.md:
```diff
@@ -0,0 +1,303 @@
+  and indexes.
+* A backfill process is distributed across all nodes in the cluster
+  which have data for the table.  This backfill will populate the shadow
+  column and shadow indexes, inserting data with the source MVCC
```
I don't think we can insert data with the source MVCC timestamp. If the timestamp is below the GC threshold then we certainly can't commit a txn at that timestamp. Also, that would require one txn per row, which is really slow. I think it's ok to use whatever the backfill chunk read time is, just like for column backfill.

mjibson commented on docs/RFCS/20180411_alter_column_type.md:
```diff
@@ -0,0 +1,303 @@
+  when a writable shadow exists for their target column or index.
+* Code that resolves a column or index name for reading should be
+  prepared to look for an activated shadow.
+* It must be possible to insert bulk, backfilled data into the LSM such
```
I'm not sure this is a thing we want to do because it would actually change history if someone was doing a read AS OF SYSTEM TIME.

madelynnblue reviewed (commented):
I think this is a great RFC. I really liked the discussion of the different classes and implementation phases.

bobvawter commented:
@knz, as I understand them, both the interleaved tables features and column families are just used to compute the KV key that the data will actually be inserted at to improve data locality.  Beyond ensuring that any shadow columns created by a type change are placed in the same column family as the source column, is there some specific problem you're thinking of?  Or is this just a topic that should appear in any RFC related to data manipulation?

---

Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

*[docs/RFCS/20180411_alter_column_type.md, line 97 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24703#-L9qBvI-6-Ari9eoxfpo-r1-97:-L9vLUIN9qfE45JRSfx3:b-q8s9yo) ([raw file](https://github.com/cockroachdb/cockroach/blob/e086e3500adce246e318bae6d60e6cbb9267a6e4/docs/RFCS/20180411_alter_column_type.md#L97)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

Does this (not changing a PK column) apply to all classes? Even `trivial`?
</blockquote></details>

My thinking is that we get non-PK columns working first to get all of the type-change machinery in place.  Once it's possible to change the PK columns for a table, we can merge the two features together, since it seems like making any non-trivial change to a PK column type is effectively the same as changing the PK to a new (shadow) column.

We could certainly support trivial type changes for PK columns as part of this RFC, except that I think we'd need to handle foreign-key references.  Let's say the user is doing `STRING -> BYTES` in table A and table B has an FK constraint.  Does that `STRING` column in table B also need to have a cascaded type change to `BYTES`?

---

*[docs/RFCS/20180411_alter_column_type.md, line 105 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24703#-L9qBvI-6-Ari9eoxfpo-r1-105:-L9vMQumEID3eAmotcod:b-chi7so) ([raw file](https://github.com/cockroachdb/cockroach/blob/e086e3500adce246e318bae6d60e6cbb9267a6e4/docs/RFCS/20180411_alter_column_type.md#L105)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

Can you define breaking change? It seems like it's anything not in the trivial class.
</blockquote></details>

Clarified the sentence.

---

*[docs/RFCS/20180411_alter_column_type.md, line 110 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24703#-L9qBvI-6-Ari9eoxfpo-r1-110:-L9vMTv91oeeAiSwzYt4:b-482z5y) ([raw file](https://github.com/cockroachdb/cockroach/blob/e086e3500adce246e318bae6d60e6cbb9267a6e4/docs/RFCS/20180411_alter_column_type.md#L110)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

What's the use of the job id? We don't return them anywhere else, and I might want to add them all at once instead of here so we can decide exactly when, why, and how we want to return job ids.
</blockquote></details>

That was left over from an earlier thought about how the two-phase commit would work.  It's not actually needed and has been removed.

---

*[docs/RFCS/20180411_alter_column_type.md, line 203 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24703#-L9qBvI-6-Ari9eoxfpo-r1-203:-L9vNhn86AIukx1vCJE0:b-su337t) ([raw file](https://github.com/cockroachdb/cockroach/blob/e086e3500adce246e318bae6d60e6cbb9267a6e4/docs/RFCS/20180411_alter_column_type.md#L203)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

I'm not sure this is a thing we want to do because it would actually change history if someone was doing a read AS OF SYSTEM TIME.
</blockquote></details>

The scenario that I want to avoid looks like:

* Start reading a range to perform the conversion.
* The user issues a write to a row within that range.
* Stomp on the user's write with the backfilled data.

It would be fine, I think to retry that range if we detect a conflict; is there plumbing that already makes this possible?

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24703#-:-L9vOwZe93jZZjTZNWE7:baxiquz)*
<!-- Sent from Reviewable.io -->

bobvawter commented:
Another concern I just thought of is range thrash: since the shadow column is going to duplicate the source column and we want to be able to have them both live in the KV layer at the same time, our encoded column families are going to get (temporarily) larger.  This could cause extra range-splitting to happen, right?  Is this going to be a problem?

---

Review status: 0 of 1 files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24703#-:-L9vWSxc8c73eZ4IK9T-:byb6cbr)*
<!-- Sent from Reviewable.io -->

knz commented:
Regarding interleaved tables: the RFC should discuss whether changing the column type on the parent should be blocked (error) or accepted (and propagate automatically to the children tables).

vivekmenezes commented on docs/RFCS/20180411_alter_column_type.md:
```diff
@@ -0,0 +1,429 @@
+run `COMPLETE` before the conversion job has finished; it is simply a
+signal to indicate that the SQL consumers are prepared to handle the new
+column type.
+
```
good thinking about this. I'm not sure if we need to do it in a first implementation though.

vivekmenezes commented on docs/RFCS/20180411_alter_column_type.md:
```diff
@@ -0,0 +1,429 @@
+retired. The shadow column will be anonymous and generally invisible to
+the user until it is swapped in. Shadow columns will be created in the
+same column family as the source column.
+
```
nice summary!

vivekmenezes commented on docs/RFCS/20180411_alter_column_type.md:
```diff
@@ -0,0 +1,429 @@
+  which have data for the table. This backfill will populate the shadow
+  column and shadow indexes, inserting data with the source MVCC
+  timestamp. The backfilled sstables will be inserted into the LSM such
+  that any shadowed write will have priority.
```
you can say here that backfilling this column is similar to how we backfill a computed column today

vivekmenezes commented on docs/RFCS/20180411_alter_column_type.md:
```diff
@@ -0,0 +1,429 @@
+the target type, the `INSERT` should fail with an error message such as
+`column user_id is being converted from STRING to UUID. The value
+'foobar' cannot be converted to UUID.`
+
```
The biggest complexity in this change is introducing the notion of shadow column everywhere in the run time code. Its quite similar to a computed column with the same name, I'd like us try to hide the name from the shadow column until it has been promoted, so that we don't have to make the assumption everywhere in the code that a duplicate column can exist. perhaps we should leave its name unspecified until the promotion happens.

vivekmenezes replied:
@justinj might have some suggestions here.

vivekmenezes reviewed (commented):
Thanks for putting this together. It is very well thought through especially for someone new to our system. Thank you!

bobvawter commented:
Review status: 0 of 1 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[docs/RFCS/20180411_alter_column_type.md, line 97 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24703#-L9qBvI-6-Ari9eoxfpo-r1-97:-LAYVqN-7vMsC4L0HjMu:brax6gm) ([raw file](https://github.com/cockroachdb/cockroach/blob/e086e3500adce246e318bae6d60e6cbb9267a6e4/docs/RFCS/20180411_alter_column_type.md#L97)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

Good point. Sounds like we should stick with no PK and no FK for now.
</blockquote></details>

Added a paragraph in the rationale section summarizing this.

---

*[docs/RFCS/20180411_alter_column_type.md, line 183 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24703#-L9qBvI-6-Ari9eoxfpo-r1-183:-LAYX1kB3p1_QobDX0i-:brwu6ho) ([raw file](https://github.com/cockroachdb/cockroach/blob/e086e3500adce246e318bae6d60e6cbb9267a6e4/docs/RFCS/20180411_alter_column_type.md#L183)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

I don't think we can insert data with the source MVCC timestamp. If the timestamp is below the GC threshold then we certainly can't commit a txn at that timestamp. Also, that would require one txn per row, which is really slow. I think it's ok to use whatever the backfill chunk read time is, just like for column backfill.
</blockquote></details>

Clarified this section to describe the ideal condition, pending discussion with DT and core-team about how we can slide sstable updates "underneath" and pre-existing writes to a particular KV key.

---

*[docs/RFCS/20180411_alter_column_type.md, line 203 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24703#-L9qBvI-6-Ari9eoxfpo-r1-203:-LAYXS6Z-aNg8CFJaLA7:b-uv5hvr) ([raw file](https://github.com/cockroachdb/cockroach/blob/e086e3500adce246e318bae6d60e6cbb9267a6e4/docs/RFCS/20180411_alter_column_type.md#L203)):*
<details><summary><i>Previously, bobvawter (Bob Vawter) wrote…</i></summary><blockquote>

The scenario that I want to avoid looks like:

* Start reading a range to perform the conversion.
* The user issues a write to a row within that range.
* Stomp on the user's write with the backfilled data.

It would be fine, I think to retry that range if we detect a conflict; is there plumbing that already makes this possible?
</blockquote></details>

Moved this into the unresolved-questions section.

---

*[docs/RFCS/20180411_alter_column_type.md, line 107 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24703#-L9qBvI-6-Ari9eoxfpo-r2-107:-LAYUizSEWzf-irqOSi_:bfpocdt) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2e2b8049e1fdee6809814b6ff81c846486d1afe/docs/RFCS/20180411_alter_column_type.md#L107)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

Are you planning on running some sort of conversion for invalid data? What does postgres do?
</blockquote></details>

Added details.

---

*[docs/RFCS/20180411_alter_column_type.md, line 176 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24703#-L9qBvI-6-Ari9eoxfpo-r2-176:-LAYVyMG6hZ-b_Ak3sZV:bm5bwrk) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2e2b8049e1fdee6809814b6ff81c846486d1afe/docs/RFCS/20180411_alter_column_type.md#L176)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

good thinking about this. I'm not sure if we need to do it in a first implementation though.
</blockquote></details>

It's not going to be necessary for a trivial conversion, since those are "instantaneous", but we'll want a two-phase setup for validated changes.

---

*[docs/RFCS/20180411_alter_column_type.md, line 204 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24703#-L9qBvI-6-Ari9eoxfpo-r2-204:-LAYW8zy1Fu3MB9uNrYF:b-ucfkn4) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2e2b8049e1fdee6809814b6ff81c846486d1afe/docs/RFCS/20180411_alter_column_type.md#L204)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

nice summary!
</blockquote></details>

Thanks.

---

*[docs/RFCS/20180411_alter_column_type.md, line 268 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24703#-L9qBvI-6-Ari9eoxfpo-r2-268:-LAYXfsc7HL-WJ7KhpTC:b-9nch0h) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2e2b8049e1fdee6809814b6ff81c846486d1afe/docs/RFCS/20180411_alter_column_type.md#L268)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

@justinj might have some suggestions here.
</blockquote></details>

Clarified the sentence in which shadow columns are introduced that they will be anonymous.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24703)*
<!-- Sent from Reviewable.io -->

bobvawter commented:
@mjibson, @knz I'd like to get to an lgtm to land this as an in-progress rfc and start laying down code.  Thanks.

---

Review status: 0 of 1 files reviewed at latest revision, 10 unresolved discussions.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24703#-:-LAYYsS51HBWfTGbYMGJ:besacfk)*
<!-- Sent from Reviewable.io -->

a-robinson commented:
Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[docs/RFCS/20180411_alter_column_type.md, line 163 at r3](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24703#-LAZRsl_6mcX2fyWahJO:-LAZRsl_6mcX2fyWahJP:b-qoaz1u) ([raw file](https://github.com/cockroachdb/cockroach/blob/76573862a5f14e9b43ddb318a7496bd5e1fd7706/docs/RFCS/20180411_alter_column_type.md#L163)):*
> ```Markdown
> -- Wait for the job to finish, stop the application, swap the column type.
> -- This command would block, returning any output from the conversion job.
> -- Any reads that happen-after this command will see the new type, while any
> ```

What happens to reads that are submitted to cockroach while the `COMPLETE` is blocking? Do they block?

---

*[docs/RFCS/20180411_alter_column_type.md, line 189 at r3](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24703#-LAZSLEjBhiDiFOVmQrC:-LAZSLEk1cNdPx9IL-A1:bdgrhn) ([raw file](https://github.com/cockroachdb/cockroach/blob/76573862a5f14e9b43ddb318a7496bd5e1fd7706/docs/RFCS/20180411_alter_column_type.md#L189)):*
<details><summary><i alt="summary">Quoted 5 lines of code…</i></summary>

> ```
> Columns that are the target of a primary or foreign key constraint will
> not be supported under this RFC. It could be possible to support foreign
> keys in the future by cascading the type change, but this would greatly
> increase the amount of coordination required to effect the total schema
> change.
> ```
</details>


This could also break views in certain circumstances. We either need to disallow changing the types of columns used in views, or do some hard thinking about how type changes affect them. @knz is the expert here.

---

*[docs/RFCS/20180411_alter_column_type.md, line 383 at r3](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24703#-LAZU0864lpxRsFGTJIx:-LAZU0864lpxRsFGTJIy:b-2rjxul) ([raw file](https://github.com/cockroachdb/cockroach/blob/76573862a5f14e9b43ddb318a7496bd5e1fd7706/docs/RFCS/20180411_alter_column_type.md#L383)):*
> ```Markdown
> since a copy-and-rename operation can be performed or users can write
> their own migration code. This is, however, a sub-optimal user
> experience and will diminish compatibility with ORM or other tooling
> ```

Nit that you don't have to bother fixing, but saying that it "will diminish compatibility" is much weaker than saying that it does diminish compatibility and calling out at least one specific important ORM/tool that is hurt by it.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24703)*
<!-- Sent from Reviewable.io -->

mjibson commented:
LGTM modulo Alex's comments.

---

Review status: all files reviewed at latest revision, 13 unresolved discussions.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24703#-:-LAlC1gz5JLKugf5gMmi:b-igqj21)*
<!-- Sent from Reviewable.io -->

knz commented:
Thanks for pinging me. Overall I'm happy about most of the details addressed but there's still one major point open. See below.

---

Review status: all files reviewed at latest revision, 13 unresolved discussions.

---

*[docs/RFCS/20180411_alter_column_type.md, line 144 at r3](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24703#-LAlVGoc1fX9qxc5DmJB:-LAlVGoc1fX9qxc5DmJC:be3gceq) ([raw file](https://github.com/cockroachdb/cockroach/blob/76573862a5f14e9b43ddb318a7496bd5e1fd7706/docs/RFCS/20180411_alter_column_type.md#L144)):*
On second read I'm afraid this DEFERRED/COMPLETE business won't do.
Remember, we're investing in this work also partly for compatibility with PostgreSQL clients. It is thus rather important that we invest sufficient time and attention in reusing PostgreSQL's SQL syntax/semantics to achieve this. 
1) pg's syntax for ALTER TABLE SET TYPE does not include DEFERRED/COMPLETE
2) we want to introduce compatibility with existing clients/apps which already use ALTER TABLE SET TYPE. These do not know about DEFERRED/COMPLETE.

So where does this leave us.

- I think there should be an alternative to the mechanisms discussed here which does not require these keywords.
- I have made a first comment about this in here: https://github.com/cockroachdb/cockroach/pull/24703#issuecomment-380576455

> about your deferred/completed approach with jobs. It seems orthogonal to this RFC: the mechanism could be equally well applied to any of our other asynchronous schema changes. If you want to argue it is necessary for "alter table set type", then I'd recommend starting this work separately, and complete it before you resume work on this feature.

I would like to see your reaction to this. It seems to me that you could issue the ALTER TABLE SET TYPE like any of our other async schema changes, ie. queue the mutation and process it upon txn commit (like CREATE INDEX, ALTER TABLE ADD COLUMN, etc). If you were to do that, then we get the compatibility with pg clients "for free". 

The mechanism you are proposing here (DEFERRED/COMPLETE) could be considered as well as a CockroachDB-extension, and made available to any ALTER not just this one.

---

*[docs/RFCS/20180411_alter_column_type.md, line 189 at r3](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24703#-LAZSLEjBhiDiFOVmQrC:-LAlWbYJ6X8N2ESPpawK:bsfbvr7) ([raw file](https://github.com/cockroachdb/cockroach/blob/76573862a5f14e9b43ddb318a7496bd5e1fd7706/docs/RFCS/20180411_alter_column_type.md#L189)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

> ```
> Columns that are the target of a primary or foreign key constraint will
> not be supported under this RFC. It could be possible to support foreign
> keys in the future by cascading the type change, but this would greatly
> increase the amount of coordination required to effect the total schema
> change.
> ```

This could also break views in certain circumstances. We either need to disallow changing the types of columns used in views, or do some hard thinking about how type changes affect them. @knz is the expert here.
</blockquote></details>

Yes it's rather simple really. The RFC must make it clear that changing the type of a column depended on by a view should be simply disallowed, there is nothing else we can do. 
The DependedOnBy field on the descriptor indicates which column IDs are "locked" in that way.

---

*[docs/RFCS/20180411_alter_column_type.md, line 291 at r3](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24703#-LAlXHVH0-qkNx3_7Ww8:-LAlXHVJ-KKp5qiyWvj1:b-cskxz0) ([raw file](https://github.com/cockroachdb/cockroach/blob/76573862a5f14e9b43ddb318a7496bd5e1fd7706/docs/RFCS/20180411_alter_column_type.md#L291)):*
> ```Markdown
> cast, it will be marked as `(extension)`.
> 
> | From                | To                      | Notes               |
> ```

This table must be extended to include TIMESTAMPTZ and TIMETZ

I'd also like to make it clear in the Notes which conversion applies. It's not clear to me what it means when the notes column is empty.

---

*[docs/RFCS/20180411_alter_column_type.md, line 337 at r3](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24703#-LAlXhOD9GvukYYiOBdA:-LAlXhOD9GvukYYiOBdB:b-7p20e) ([raw file](https://github.com/cockroachdb/cockroach/blob/76573862a5f14e9b43ddb318a7496bd5e1fd7706/docs/RFCS/20180411_alter_column_type.md#L337)):*
> ```Markdown
> |                     | `INTERVAL`              |                     |
> |                     | `DECIMAL`               |                     |
> |                     | `FLOAT`                 |                     |
> ```

float is listed twice

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24703#-:-LAlXnTgBk8qaTg9S7X1:b-33x4ma)*
<!-- Sent from Reviewable.io -->

knz commented:
Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 16 unresolved discussions.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24703)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented on docs/RFCS/20180411_alter_column_type.md:
```diff
@@ -0,0 +1,447 @@
+ALTER TABLE rpc_logs
+  ALTER COLUMN caller_ip
+  SET DATA TYPE INET
+  DEFERRED
```
I think this command should block until the the entire type conversion has happened (with backfill), preparing the table for the quick execution of COMPLETE below.

vivekmenezes commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 13 unresolved discussions.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24703#-:-LAmtirTAdbAKM4UxUF0:bnfp4nl)*
<!-- Sent from Reviewable.io -->

bobvawter commented:
@knz I've responded to your concerns about new keywords.  I'm open to finding some other mechanism for controlling a two-phase schema change, but I don't think that using SQL transactions for control is the right approach.  If you still object to the new keywords, let's set up a call with any other interested parties to figure out a third option to get to closure faster.

---

Review status: all files reviewed at latest revision, 13 unresolved discussions.

---

*[docs/RFCS/20180411_alter_column_type.md, line 101 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24703#-L9qBvI-6-Ari9eoxfpo-r2-101:-LAmh2IG8QS48Zj0qeiJ:bhy9ot6) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2e2b8049e1fdee6809814b6ff81c846486d1afe/docs/RFCS/20180411_alter_column_type.md#L101)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

you can go ahead and implement these type conversions ^
</blockquote></details>

Will do.

---

*[docs/RFCS/20180411_alter_column_type.md, line 221 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24703#-L9qBvI-6-Ari9eoxfpo-r2-221:-LAmrUMw2SSZiuEMKrKj:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2e2b8049e1fdee6809814b6ff81c846486d1afe/docs/RFCS/20180411_alter_column_type.md#L221)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

you can say here that backfilling this column is similar to how we backfill a computed column today
</blockquote></details>

Done.

---

*[docs/RFCS/20180411_alter_column_type.md, line 350 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24703#-LAK2cJp2b9Y7q7iSpPr:-LAmrWYe2dJtARGOrZ51:bt2vr1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/a2e2b8049e1fdee6809814b6ff81c846486d1afe/docs/RFCS/20180411_alter_column_type.md#L350)):*
<details><summary><i>Previously, mjibson (Matt Jibson) wrote…</i></summary><blockquote>

I thought the collation key affected the on-disk bytes. Does it not?
</blockquote></details>

Fixed.

---

*[docs/RFCS/20180411_alter_column_type.md, line 144 at r3](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24703#-LAlVGoc1fX9qxc5DmJB:-LAmhK4mB-pUE18byv81:b5avhn9) ([raw file](https://github.com/cockroachdb/cockroach/blob/76573862a5f14e9b43ddb318a7496bd5e1fd7706/docs/RFCS/20180411_alter_column_type.md#L144)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

One idea is to introduce the notion of BACKFILL. So without using this keyword cockroach can behave like pg and run all the individual steps for the schema change. With the BACKFILL keyword, it completes all the steps until the backfill is complete. The BACKFILL command can be followed by the plain command that is speeded up because the BACKFILL is complete
</blockquote></details>

@knz: I may have mis-read the above, but would your approach look like

```
BEGIN
ALTER TABLE foo ...
COMMIT
```

Where the `ALTER TABLE` command blocks until the backfill is complete and the `COMMIT` actually makes the change visible?

I don't think that using transactions are the right way to control a two-phase schema change for the following reasons:

* As I understand it, PostgreSQL [locks the table](https://www.postgresql.org/docs/current/static/sql-altertable.html)  when executing a schema change.  This is something that's antithetical to Cockroach's approach, so we're never going to match pg's behavior 100%.
* It is conceivable that for a large data set, the amount of time it takes to perform a validation or a backfill would exceed the amount of time that a single transaction could (or should) be held open for.
* The schema change initiated within a transaction is necessarily going to have observable side-effects for writers in order to make a cast-type backfill work well in the face of non-trivial amounts of existing data to convert.  Either we disallow writes to the affected column in other transactions or impose the proposed write-duplication behavior (with additional value constraints).

The new keywords are optional, and if you don't use `ALTER TABLE ... DEFERRED`, the command will block until the schema change is visible.  This should satisfy most simple use cases.

I do agree with you that we should extend explicit schema control to all `ALTER TABLE` statements.  Once we've been able to get some experience with the new control mechanism and ensure that it works well, we can then upgrade the existing `ALTER TABLE` flows to do the same.

If you find this argument in favor of new keywords satisfactory, I'll clean it up and add it to the rationale section in the doc.

@vivekmenezes: What if we introduce a new transaction kind like `BEGIN BACKFILL TRANSACTION`?  This would disallow any DML statements and allow you to queue up multiple type changes for a single table, while still only performing a single backfill operation.

---

*[docs/RFCS/20180411_alter_column_type.md, line 157 at r3](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24703#-L9qBvI-6-Ari9eoxfpo-r3-157:-LAmqY3U6B9h_bhOgXua:b-4jchym) ([raw file](https://github.com/cockroachdb/cockroach/blob/76573862a5f14e9b43ddb318a7496bd5e1fd7706/docs/RFCS/20180411_alter_column_type.md#L157)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

I think this command should block until the the entire type conversion has happened (with backfill), preparing the table for the quick execution of COMPLETE below.
</blockquote></details>

Makes sense.  The job-id is really just an implementation detail that the user can choose to investigate if they actually care about it.

---

*[docs/RFCS/20180411_alter_column_type.md, line 163 at r3](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24703#-LAZRsl_6mcX2fyWahJO:-LAm_zTg3rOsisb2Ijr7:baqzt2u) ([raw file](https://github.com/cockroachdb/cockroach/blob/76573862a5f14e9b43ddb318a7496bd5e1fd7706/docs/RFCS/20180411_alter_column_type.md#L163)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

What happens to reads that are submitted to cockroach while the `COMPLETE` is blocking? Do they block?
</blockquote></details>

It's documented just below here, but issuing `COMPLETE` just lets the database know that it's ok to perform the column swap once everything is in place to do so.

---

*[docs/RFCS/20180411_alter_column_type.md, line 189 at r3](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24703#-LAZSLEjBhiDiFOVmQrC:-LAmad-E6mCpTT-qfTVY:b-yfvnwb) ([raw file](https://github.com/cockroachdb/cockroach/blob/76573862a5f14e9b43ddb318a7496bd5e1fd7706/docs/RFCS/20180411_alter_column_type.md#L189)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Yes it's rather simple really. The RFC must make it clear that changing the type of a column depended on by a view should be simply disallowed, there is nothing else we can do. 
The DependedOnBy field on the descriptor indicates which column IDs are "locked" in that way.
</blockquote></details>

Agreed.  Columns that are involved in a view will be explicitly de-scoped for the first round.  I suppose handling views is going to be similar to handling columns involved in FK: cascade the type change through the schema.  Fortunately, for the purposes of schema-change, views are dematerialized so we don't need another backfill.

---

*[docs/RFCS/20180411_alter_column_type.md, line 291 at r3](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24703#-LAlXHVH0-qkNx3_7Ww8:-LAmdDZw7h_WoCX1CraN:b-qwks30) ([raw file](https://github.com/cockroachdb/cockroach/blob/76573862a5f14e9b43ddb318a7496bd5e1fd7706/docs/RFCS/20180411_alter_column_type.md#L291)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

This table must be extended to include TIMESTAMPTZ and TIMETZ

I'd also like to make it clear in the Notes which conversion applies. It's not clear to me what it means when the notes column is empty.
</blockquote></details>

Added missing types and a clarifying sentence above the table explaining that cast-type is the default.  When I had "Cast" listed in every note cell, the table was visually cluttered without adding any real value.

---

*[docs/RFCS/20180411_alter_column_type.md, line 337 at r3](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24703#-LAlXhOD9GvukYYiOBdA:-LAmfsSSClG6GtG4ckTK:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/76573862a5f14e9b43ddb318a7496bd5e1fd7706/docs/RFCS/20180411_alter_column_type.md#L337)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

float is listed twice
</blockquote></details>

Done.

---

*[docs/RFCS/20180411_alter_column_type.md, line 383 at r3](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24703#-LAZU0864lpxRsFGTJIx:-LAmgvKs-7WG3Ol2V32-:bhex00a) ([raw file](https://github.com/cockroachdb/cockroach/blob/76573862a5f14e9b43ddb318a7496bd5e1fd7706/docs/RFCS/20180411_alter_column_type.md#L383)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

Nit that you don't have to bother fixing, but saying that it "will diminish compatibility" is much weaker than saying that it does diminish compatibility and calling out at least one specific important ORM/tool that is hurt by it.
</blockquote></details>

Added links to some GitHub issues.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24703#-:-LAmrsr32_G-pmHNfx5z:b-dk6er1)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Review status: all files reviewed at latest revision, 11 unresolved discussions.

---

*[docs/RFCS/20180411_alter_column_type.md, line 144 at r3](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24703#-LAlVGoc1fX9qxc5DmJB:-LAmx67B8TwFRXSjIDPI:bku7vti) ([raw file](https://github.com/cockroachdb/cockroach/blob/76573862a5f14e9b43ddb318a7496bd5e1fd7706/docs/RFCS/20180411_alter_column_type.md#L144)):*
<details><summary><i>Previously, bobvawter (Bob Vawter) wrote…</i></summary><blockquote>

@knz: I may have mis-read the above, but would your approach look like

```
BEGIN
ALTER TABLE foo ...
COMMIT
```

Where the `ALTER TABLE` command blocks until the backfill is complete and the `COMMIT` actually makes the change visible?

I don't think that using transactions are the right way to control a two-phase schema change for the following reasons:

* As I understand it, PostgreSQL [locks the table](https://www.postgresql.org/docs/current/static/sql-altertable.html)  when executing a schema change.  This is something that's antithetical to Cockroach's approach, so we're never going to match pg's behavior 100%.
* It is conceivable that for a large data set, the amount of time it takes to perform a validation or a backfill would exceed the amount of time that a single transaction could (or should) be held open for.
* The schema change initiated within a transaction is necessarily going to have observable side-effects for writers in order to make a cast-type backfill work well in the face of non-trivial amounts of existing data to convert.  Either we disallow writes to the affected column in other transactions or impose the proposed write-duplication behavior (with additional value constraints).

The new keywords are optional, and if you don't use `ALTER TABLE ... DEFERRED`, the command will block until the schema change is visible.  This should satisfy most simple use cases.

I do agree with you that we should extend explicit schema control to all `ALTER TABLE` statements.  Once we've been able to get some experience with the new control mechanism and ensure that it works well, we can then upgrade the existing `ALTER TABLE` flows to do the same.

If you find this argument in favor of new keywords satisfactory, I'll clean it up and add it to the rationale section in the doc.

@vivekmenezes: What if we introduce a new transaction kind like `BEGIN BACKFILL TRANSACTION`?  This would disallow any DML statements and allow you to queue up multiple type changes for a single table, while still only performing a single backfill operation.
</blockquote></details>

This seems like something specific to this command so I don't think it's a good idea to add something to something as general as BEGIN TRANSACTION.  Just looking at the postgres docs I can't finding a keyword appropriate for what we're doing here. DEFERRED means something else altogether

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24703)*
<!-- Sent from Reviewable.io -->

bobvawter commented:
Based on a conversation with @knz, @mjibson, and @dt we'll take a different approach when the user needs control over exactly when a type change will become visible to clients that cannot support both the old and new column types concurrently.  The rough approach looks like the following:

* Add a new computed column from the active, original column that performs the desired type conversion.
* Wait for the computed column to backfill.
* Atomically swap the source and computed column, remove the computed-ness, and optionally drop the original column.

A regular `ALTER COLUMN TYPE` command is more or less a wrapper around the above using a hidden column and performing the swap when the backfill is ready.  This will give us the desired user experience of directed schema evolution and provide an independently-useful feature of being able to drop the computed nature of a column.  We'll also make sure that the above workflow is highly visible in the documentation.

---

Review status: 0 of 1 files reviewed at latest revision, 11 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24703#-:-LAnfnMjA2tE1hRIPJb-:bsqawoh)*
<!-- Sent from Reviewable.io -->

bobvawter committed:
RFCS: Alter column type

This RFC describes how we will support altering column types in an online
fashion.

Release note: None

bobvawter commented:
PTAL:  I've pushed a new version of the RFC that eliminates the explicit syntax for controlling the two-phase type change for one that builds on top of being able to add new computed columns.

mjibson commented on docs/RFCS/20180411_alter_column_type.md:
```diff
@@ -0,0 +1,386 @@
+  * The hidden flag is removed from the shadow column.
+  * The source column is dropped.
+  * The shadow column is renamed.
+  * Any indexes on the original column are dropped.
```
This will break any queries that specify the index by name in a query. I think that's ok though, and we can document it. I'm just thinking through this.

knz replied:
Indeed. Also indexes have a DependedOnBy populated by views, this needs to be taken into account too.

bobvawter replied:
We should be able to rename the shadow indexes while we're mucking around in those data structures, right?

knz replied:
Yes indeed. renaming the shadow index(es) to be the original index name(s) should do the trick.

knz commented on docs/RFCS/20180411_alter_column_type.md:
```diff
@@ -0,0 +1,387 @@
+
+## Views
+
+Columns that are depended upon by a `CREATE VIEW` are out of scope for
```
Clarify that the changes will ensure that a client receives an error if they attempt to chang ethe type of a column depended on by a view.

knz replied:
Also indexes as a whole can be depended on by something else (not just FKs). 

Try this: "In general any time there is a structural dependency on a column or an index, the alter proposed here will bail out with an error. Supporting these cases is left to later work."

knz commented on docs/RFCS/20180411_alter_column_type.md:
```diff
@@ -0,0 +1,387 @@
+  the source column's family, with the expression provided by the user
+  via `USING expression` or inferred for cast-type conversions.
+  * If there are indexes that refer to the column, additional
+    shadow-indexes are created.
```
Note for the implementation: be mindful that cockroachdb currently supports (for better or for worse) multiple indexes over the same set of columns, but with different names. Perhaps we shouldn't do that, but we can. There must be as many shadow indexes created as there were indexes originally, and the atomic rename described below must apply separately to all of them.

knz commented on docs/RFCS/20180411_alter_column_type.md:
```diff
@@ -0,0 +1,387 @@
+In the case where a column-type conversion (e.g. `STRING -> UUID`) is in
+process and a user attempts to insert data that is not compatible with
+the target type, the `INSERT` should fail with an error message such as
+`column user_id is being converted from STRING to UUID. The value
```
if you reuse the computed column mechanism, it may not be practical / elegant to change the error message. Please check what would currently happen with a computed column, and see if the error message is acceptable from a UX perspective. If it's not, then this RFC should outline how you plan to extend the code for error handling, and/or if it would be possible to use the same error messages for both regular computed columns and type substitutions as described here. I would prefer to avoid weird boolean flags throughout multiple layers of abstraction just to get the right error message.

knz commented on docs/RFCS/20180411_alter_column_type.md:
```diff
@@ -0,0 +1,387 @@
+
+## Job control
+
+The type change will be modeled as a bulk-IO job, allowing it to be
```
Make a reference here of how this will align with the handling of other DDL statements, and highlight that you plan to reuse the same code paths.

(Note that not all DDL statements are currently handled as jobs — although arguably they should — so pick one that is already as example)

knz commented on docs/RFCS/20180411_alter_column_type.md:
```diff
@@ -0,0 +1,387 @@
+The element type of an array can be converted, based on the conversion
+rules above. That is, `STRING[]` could be converted to `DATE[]`.
+Additionally, it is possible to convert any type to an array of a
+convertable type, e.g. `STRING -> DATE[]`.
```
can you expand on this. What is the expected semantics of this last conversion? What does pg does in that case? What is the use case?

In general I would be rather strongly against the idea of enabling an automatic (i.e. non-assisted) column type conversion that is not otherwise possible via a cast (using `::`). This last example would be an example of that. I'd need a use case to be convinced otherwise.

knz reviewed (commented):
Thanks for all the improvements already.

vivekmenezes commented on docs/RFCS/20180411_alter_column_type.md:
```diff
@@ -0,0 +1,387 @@
+Once the type-change work has been completed, the new type becomes
+effective and the old data can be retired. We allow only a single
+column-type change to be in progress for any given table at a time;
+multiple type changes will be queued and executed sequentially.
```
you can add, "although this design doesn't preclude simultaneous execution of multiple type changes"

vivekmenezes reviewed (commented):
Looks Good. I'd appreciate you provide an example of how a user will control a manual two phase column type conversion. Thanks!

bobvawter commented:
Review status: 0 of 1 files reviewed at latest revision, 18 unresolved discussions.

---

*[docs/RFCS/20180411_alter_column_type.md, line 208 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/24703#-L9qBvI-6-Ari9eoxfpo-r5-208:-LB1_Xka754ULbz629N_:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/06d67a7f93d8ac849841c223bc0ebc31a913027c/docs/RFCS/20180411_alter_column_type.md#L208)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Yes indeed. renaming the shadow index(es) to be the original index name(s) should do the trick. 
</blockquote></details>

Done.

---

*[docs/RFCS/20180411_alter_column_type.md, line 83 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/24703#-L9qBvI-6-Ari9eoxfpo-r6-83:-LB1dx5CAFd4FwUoMDm4:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a148046861d67226e4acea31d84f712cbab26d44/docs/RFCS/20180411_alter_column_type.md#L83)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

you can add, "although this design doesn't preclude simultaneous execution of multiple type changes"
</blockquote></details>

Done.

---

*[docs/RFCS/20180411_alter_column_type.md, line 167 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/24703#-L9qBvI-6-Ari9eoxfpo-r6-167:-LB1Zxag7R9EHQW3T64f:b2ubji6) ([raw file](https://github.com/cockroachdb/cockroach/blob/a148046861d67226e4acea31d84f712cbab26d44/docs/RFCS/20180411_alter_column_type.md#L167)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Also indexes as a whole can be depended on by something else (not just FKs). 

Try this: "In general any time there is a structural dependency on a column or an index, the alter proposed here will bail out with an error. Supporting these cases is left to later work."
</blockquote></details>

Reworked this section to talk more generally about not supporting structural dependencies other than indexes.

---

*[docs/RFCS/20180411_alter_column_type.md, line 192 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/24703#-L9qBvI-6-Ari9eoxfpo-r6-192:-LB1_PG962gZxsT7nTs2:b-yzall2) ([raw file](https://github.com/cockroachdb/cockroach/blob/a148046861d67226e4acea31d84f712cbab26d44/docs/RFCS/20180411_alter_column_type.md#L192)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Note for the implementation: be mindful that cockroachdb currently supports (for better or for worse) multiple indexes over the same set of columns, but with different names. Perhaps we shouldn't do that, but we can. There must be as many shadow indexes created as there were indexes originally, and the atomic rename described below must apply separately to all of them.
</blockquote></details>

Reworded to be more general.

---

*[docs/RFCS/20180411_alter_column_type.md, line 234 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/24703#-L9qBvI-6-Ari9eoxfpo-r6-234:-LB1a6BtD4kwsW0Hgt8O:b-emhkx7) ([raw file](https://github.com/cockroachdb/cockroach/blob/a148046861d67226e4acea31d84f712cbab26d44/docs/RFCS/20180411_alter_column_type.md#L234)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

if you reuse the computed column mechanism, it may not be practical / elegant to change the error message. Please check what would currently happen with a computed column, and see if the error message is acceptable from a UX perspective. If it's not, then this RFC should outline how you plan to extend the code for error handling, and/or if it would be possible to use the same error messages for both regular computed columns and type substitutions as described here. I would prefer to avoid weird boolean flags throughout multiple layers of abstraction just to get the right error message.
</blockquote></details>

Understood.  Updated this paragraph to discuss the alternative error message.

---

*[docs/RFCS/20180411_alter_column_type.md, line 239 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/24703#-L9qBvI-6-Ari9eoxfpo-r6-239:-LB1bly-DWAD7yKJzwml:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a148046861d67226e4acea31d84f712cbab26d44/docs/RFCS/20180411_alter_column_type.md#L239)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Make a reference here of how this will align with the handling of other DDL statements, and highlight that you plan to reuse the same code paths.

(Note that not all DDL statements are currently handled as jobs — although arguably they should — so pick one that is already as example)
</blockquote></details>

Done.

---

*[docs/RFCS/20180411_alter_column_type.md, line 338 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/24703#-L9qBvI-6-Ari9eoxfpo-r6-338:-LB1cYb8-HgJprAsIvuv:btp09vu) ([raw file](https://github.com/cockroachdb/cockroach/blob/a148046861d67226e4acea31d84f712cbab26d44/docs/RFCS/20180411_alter_column_type.md#L338)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

can you expand on this. What is the expected semantics of this last conversion? What does pg does in that case? What is the use case?

In general I would be rather strongly against the idea of enabling an automatic (i.e. non-assisted) column type conversion that is not otherwise possible via a cast (using `::`). This last example would be an example of that. I'd need a use case to be convinced otherwise.
</blockquote></details>

Agreed.  I have clarified that only `T -> T[]` can be performed implicitly.  Any other array-involved conversion `T -> S[]` or `T[] -> T` requires an explicit `USING expression`.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/24703)*
<!-- Sent from Reviewable.io -->

bobvawter commented:
PTAL: Added `EXPLAIN(STEPS)` as a way to give DBA's a way to perform preparatory work so the actual `ALTER ... COLUMN TYPE` can be almost instantaneous.

---

Review status: 0 of 1 files reviewed at latest revision, 18 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/24703#-:-LB1gfKk2mIcS7yendY4:b1hp5pg)*
<!-- Sent from Reviewable.io -->

bobvawter commented:
bors r+

knz commented:
Reviewed 1 of 1 files at r4.
Review status: 0 of 1 files reviewed at latest revision, 18 unresolved discussions, some commit checks pending.

---

*[docs/RFCS/20180411_alter_column_type.md, line 338 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/24703#-L9qBvI-6-Ari9eoxfpo-r6-338:-LBW4CUDC76m2hUFdKIA:bhskekc) ([raw file](https://github.com/cockroachdb/cockroach/blob/a148046861d67226e4acea31d84f712cbab26d44/docs/RFCS/20180411_alter_column_type.md#L338)):*
<details><summary><i>Previously, bobvawter (Bob Vawter) wrote…</i></summary><blockquote>

Agreed.  I have clarified that only `T -> T[]` can be performed implicitly.  Any other array-involved conversion `T -> S[]` or `T[] -> T` requires an explicit `USING expression`.
</blockquote></details>

I must insist, the conversion T -> T[] is 

1) not clearly motivated here 
2) not possible via casts 
3) hard to justify implementation-wise: there's nothing that guarantees it will be cheap other than currnet implementation happenstance

I would encourage you to _not_ implement this initially *at all* and instead wait until you have completed the rest of the work to decide whether it is indeed useful (and possibly leave that implementation to a follow-up project).

I could be convinced otherwise if I saw the business case, but the RFC is not presenting that yet.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/24703)*
<!-- Sent from Reviewable.io -->

bobvawter commented:
Review status: 0 of 1 files reviewed at latest revision, 18 unresolved discussions, some commit checks failed.

---

*[docs/RFCS/20180411_alter_column_type.md, line 338 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/24703#-L9qBvI-6-Ari9eoxfpo-r6-338:-LBXN7t68K_5aPqmd6Sr:bpqhuog) ([raw file](https://github.com/cockroachdb/cockroach/blob/a148046861d67226e4acea31d84f712cbab26d44/docs/RFCS/20180411_alter_column_type.md#L338)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I must insist, the conversion T -> T[] is 

1) not clearly motivated here 
2) not possible via casts 
3) hard to justify implementation-wise: there's nothing that guarantees it will be cheap other than currnet implementation happenstance

I would encourage you to _not_ implement this initially *at all* and instead wait until you have completed the rest of the work to decide whether it is indeed useful (and possibly leave that implementation to a follow-up project).

I could be convinced otherwise if I saw the business case, but the RFC is not presenting that yet.
</blockquote></details>

`T -> T[]` and `T[] -> T` must be possible if we're implementing a general type-conversion system.  I'm assuming that cast-type conversions will be implemented by substituting a default `USING` clause, which in this case would be `USING ARRAY[column_name]`.  The implicit conversion is a user convenience and doesn't affect the rest of the type system.  If it turns into an implementation problem, I'll reconsider allowing array conversions, but, for now, I remain unconvinced that it should be excluded.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/24703)*
<!-- Sent from Reviewable.io -->

knz commented:
Reviewed 1 of 1 files at r7.
Review status: all files reviewed at latest revision, 18 unresolved discussions, some commit checks failed.

---

*[docs/RFCS/20180411_alter_column_type.md, line 338 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/24703#-L9qBvI-6-Ari9eoxfpo-r6-338:-LBXSibFD-0TqrSzA5Jp:bdcamzq) ([raw file](https://github.com/cockroachdb/cockroach/blob/a148046861d67226e4acea31d84f712cbab26d44/docs/RFCS/20180411_alter_column_type.md#L338)):*
> `T -> T[]` and `T[] -> T` must be possible if we're implementing a general type-conversion system.

1) Why?

We do not support conversion of UUID to/from FLOAT. We do not support conversion of INET to/from INTERVAL. The reason why we do not support them is that they are non-sensical.

There are plenty of non-sensical conversions.

I posit that converting T <-> T[] is also non-sensical. I am willing to change my mind, but you'd need to explain. 

2) What is a "general type-conversion system"?

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/24703)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: all files reviewed at latest revision, 18 unresolved discussions, some commit checks failed.

---

*[docs/RFCS/20180411_alter_column_type.md, line 338 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/24703#-L9qBvI-6-Ari9eoxfpo-r6-338:-LBXdC6R31VpejvBktuZ:br10zc) ([raw file](https://github.com/cockroachdb/cockroach/blob/a148046861d67226e4acea31d84f712cbab26d44/docs/RFCS/20180411_alter_column_type.md#L338)):*
<details><summary><i>Previously, bobvawter (Bob Vawter) wrote…</i></summary><blockquote>

You know the saying about how there's only three numbers that matter: 0, 1, ∞, and everything else is a special case?  The way I see it, you should be able to have 0 of something (a never-created or dropped column), 1 of something (a column of type `T`) or an infinite number of something (a column of type `T[]`).

Right now, you can achieve the following conversions:
* `0 -> 1` is `ALTER TABLE ADD COLUMN t T`
* `0 -> ∞` is `ALTER TABLE ADD COLUMN t T[]`
* `1, ∞ -> 0` is `ALTER TABLE DROP COLUMN t`

I want to solve for the missing conversion of `1 <--> ∞` by allowing the type of a column to be changed from `T -> T[]` or `T[] ->T`.
</blockquote></details>

- T -> T[]:  `ALTER COLUMN x SET TYPE int[] USING ARRAY[x]`
- T[] -> T: `ALTER COLUMN x SET TYPE int USING x[1]`

What's missing?

In your opinion, why does PostgreSQL not allow writing `123::int::int[]`? 

In your opinion, why does PostgreSQL now allow writing `array[1,2,3]::int`?

Also what does T[] -> T mean when there's more than one element?

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/24703)*
<!-- Sent from Reviewable.io -->

bobvawter commented:
Review status: all files reviewed at latest revision, 18 unresolved discussions, some commit checks failed.

---

*[docs/RFCS/20180411_alter_column_type.md, line 338 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/24703#-L9qBvI-6-Ari9eoxfpo-r6-338:-LBXkIFpEcisX4-j_cVi:bphh5to) ([raw file](https://github.com/cockroachdb/cockroach/blob/a148046861d67226e4acea31d84f712cbab26d44/docs/RFCS/20180411_alter_column_type.md#L338)):*
> T -> T[]:  ALTER COLUMN x SET TYPE int[] USING ARRAY[x]
> T[] -> T: ALTER COLUMN x SET TYPE int USING x[1]

Exactly.  I'm not sure what we're discussing.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/24703)*
<!-- Sent from Reviewable.io -->

knz commented:
Review status: all files reviewed at latest revision, 18 unresolved discussions, some commit checks failed.

---

*[docs/RFCS/20180411_alter_column_type.md, line 338 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/24703#-L9qBvI-6-Ari9eoxfpo-r6-338:-LBXu_X59dqFsmnJWWwo:b-hny44a) ([raw file](https://github.com/cockroachdb/cockroach/blob/a148046861d67226e4acea31d84f712cbab26d44/docs/RFCS/20180411_alter_column_type.md#L338)):*
Look at the first sentence in the paragraph:

> Any given type `T` may be **implicitly** converted to `T[]`.

(emphasis mine)

If you were to rephrase this as "Any given type `T` may be *explicitly* converted to `T[]` with USING."

Then we'd be in agreement.

---


*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/24703)*
<!-- Sent from Reviewable.io -->

bobvawter commented:
@knz Updated doc so that all array-involved conversions require an explicit `USING expression`.

bobvawter commented:
bors r+
